### PR TITLE
Stabilize retry timeout error with customer error on activity startToCloseTimeout

### PR DESCRIPTION
**Pull request by longquanzheng**
for https://github.com/uber/cadence/issues/2627

tested locally:
```
2019-11-11T11:26:12.191-0800	ERROR	helloworld/helloworld_workflow.go:49	Activity failed 2.	{"Domain": "samples-domain", "TaskList": "helloWorldGroup", "WorkerID": "60286@longer-C02V60N3HTDG@helloWorldGroup", "WorkflowType": "main.Workflow", "WorkflowID": "helloworld_47fb064f-a9cf-4002-8ce6-0eb8b0badb92", "RunID": "cb1fbf38-8998-46a7-8107-c35bbbfe6164", "error": "longer error"}
main.Workflow
```

### CODE REVIEW

File: idl/github.com/uber/cadence/shared.thrift [hunk 1]
```diff
@@ -648,6 +648,10 @@ struct ActivityTaskTimedOutEventAttributes {
   10: optional i64 (js.type = "Long") scheduledEventId
   20: optional i64 (js.type = "Long") startedEventId
   30: optional TimeoutType timeoutType
+  // For retry activity, it may have a failure before timeout. It's important to keep those information for debug.
+  // Client can also provide the info for making next decision
+  40: optional string lastFailuireReason
```
Comment (vancexu) [inline idl/github.com/uber/cadence/shared.thrift]: Typo failure?

Comment (longquanzheng) [reply idl/github.com/uber/cadence/shared.thrift]: good catch, I will fix it

**Comment by longquanzheng** [comment]
> Maybe worth to add an integration test?

sure

File: internal/internal_event_handlers.go [hunk 2]
```diff
@@ -970,9 +970,17 @@ func (weh *workflowExecutionEventHandlerImpl) handleActivityTaskTimedOut(event *
...
 
+	var err error
 	attributes := event.ActivityTaskTimedOutEventAttributes
-	details := newEncodedValues(attributes.Details, weh.GetDataConverter())
-	err := NewTimeoutError(attributes.GetTimeoutType(), details)
+	if len(attributes.GetLastFailureReason()) > 0 {
+		// When retry activity timeout, it is possible that previous attempts got other customer timeout errors.
+		// To stabilize the error type, we always return the customer error.
+		// See more details of background: https://github.com/uber/cadence/issues/2627
+		err = constructError(attributes.GetLastFailureReason(), attributes.LastFailureDetails, weh.GetDataConverter())
```
Comment (yiminc) [inline internal/internal_event_handlers.go]: the last error reason could be timeout as well, could you test that as well?

Comment (longquanzheng) [reply internal/internal_event_handlers.go]: addressed offline. Our test case will cover it.

**Pull request merged**